### PR TITLE
add LD eventsource to our sdk

### PR DESCRIPTION
**Pull request by elliotCamblor**
Copied the LD code into an eventsource directory. The main difference is that we added `EventHandler.kt` which is a class that implements the EventHandler interface

### CODE REVIEW

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt [hunk 1]
```diff
@@ -0,0 +1,71 @@
+package com.devcycle.sdk.android.eventsource
+
+import timber.log.Timber
+import com.devcycle.sdk.android.eventsource.EventHandler
+
+class EventHandle: EventHandler {
```
Comment (nsalamad) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt:6]: should we change this to `EventHandler`

Comment (elliotCamblor) [reply android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt:6]: It doesn't seem to like having a class with the same name as the interface. Perhaps naming it something like `DVCEventHandler` would make sense? Not sure what the convention is here

Comment (nsalamad) [reply android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt:6]: oh I see. Nvm then. Adam called the iOS equivalent just `Handler` so we could go with that

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/api/DVCClient.kt [hunk 2]
```diff
@@ -57,6 +59,9 @@ class DVCClient private constructor(
             try {
                 fetchConfig(user)
                 isInitialized.set(true)
+                println("URL ${config?.sse?.url}")
```
Comment (emmawillis) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/api/DVCClient.kt:62]: should this println be a timber log?

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+class Handler: EventHandler {
+    fun init() {}
+
+    /**
+     * EventSource calls this method when the stream connection has been opened.
+     * @throws Exception throwing an exception here will cause it to be logged and also sent to [.onError]
+     */
+    override fun onOpen() {
+        println("OPEN")
```
Comment (emmawillis) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt:14]: same with these ones - we probably only want timber right?

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+    /**
+     * EventSource calls this method when it has received a new event from the stream.
+     * @param event the event name, from the `event:` line in the stream
+     * @param messageEvent a [MessageEvent] object containing all the other event properties
+     * @throws Exception throwing an exception here will cause it to be logged and also sent to [.onError]
+     */
+    @Throws(Exception::class)
+    override fun onMessage(event: String?, messageEvent: MessageEvent?) {
+        println("BIG OLE MESSAGE $event")
```
Comment (jsalaber) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt:45]: lol

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/model/BucketedUserConfig.kt [hunk 5]
```diff
@@ -32,5 +32,7 @@ data class BucketedUserConfig internal constructor(
     @get:Schema(description = "Map of `Variable.key` to `Variable` values.")
     val variables: Map<String, Variable<Any>>? = null,
     @get:Schema(description = "Hashes `murmurhash.v3(variable.key + environment.apiKey)` of all known variable keys not contained in the `variables` object.")
-    val knownVariableKeys: List<BigDecimal>? = null
+    val knownVariableKeys: List<BigDecimal>? = null,
+    @get:Schema(description = "Hashes `murmurhash.v3(variable.key + environment.apiKey)` of all known variable keys not contained in the `variables` object.")
```
Comment (jsalaber) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/model/BucketedUserConfig.kt:36]: this description needs to be updated

**Review by jsalaber — CHANGES_REQUESTED** [review]

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt (hunk 3 above)
Comment (elliotCamblor) [reply android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt:14]: yeah, gotta clean these up

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/model/SSE.kt [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty
+import io.swagger.v3.oas.annotations.media.Schema
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+
+class SSE {
+    /**
+     * SSE connection URL, that is set by the Bucketed User Config -> SSE -> URL
+     */
+    @get:Schema(required = false, description = "flag that determines whether or not EdgeDB is enabled")
```
Comment (jsalaber) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/model/SSE.kt:13]: this needs to be updated too

**Review by jsalaber — CHANGES_REQUESTED** [review]

File: android-client-sdk/build.gradle [hunk 7]
```diff
@@ -73,6 +73,7 @@ ext {
     mockk_version = '1.12.4'
     hamcrest_version = "2.2"
     okhttp_version = "4.9.3"
+    ld_eventsource_version = "2.7.1"
```
Comment (ajwootto) [inline android-client-sdk/build.gradle:76]: should this dependency be here? it seems like you're just bundling the whole library in the SDK?

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+     *
+     *
+     * This method is *not* called if the error was already passed to a [ConnectionErrorHandler]
+     * which returned [ConnectionErrorHandler.Action.SHUTDOWN]; EventSource assumes that if you registered
+     * such a handler and made it return that value, then you do not want to handle the same error twice.
+     *
+     * @param t  a `Throwable` object
+     */
+    override fun onError(t: Throwable?) {
+        println("BIG OLE ERROR $t")
```
Comment (ajwootto) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt:65]: can't merge with test logs in here, have to assume that the code on main will be released at any time

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/model/SSE.kt [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+    @JsonProperty("url")
+    var url: String? = null
```
Comment (ajwootto) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/model/SSE.kt:15]: this should have the inactivityDelay in it as well

File: android-client-sdk/build.gradle (hunk 7 above)
Comment (elliotCamblor) [reply android-client-sdk/build.gradle:76]: no its not, I forgot to remove it

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt (hunk 8 above)
Comment (elliotCamblor) [reply android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt:65]: yep good catch

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/api/DVCApiClient.kt [hunk 10]
```diff
@@ -2,6 +2,7 @@ package com.devcycle.sdk.android.api
 
 import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
 import okhttp3.OkHttpClient
+import com.devcycle.sdk.android.eventsource.EventSource
```
Comment (ajwootto) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/api/DVCApiClient.kt:5]: remove this?

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+    /**
+     * EventSource calls this method when it has received a new event from the stream.
+     * @param event the event name, from the `event:` line in the stream
+     * @param messageEvent a [MessageEvent] object containing all the other event properties
+     * @throws Exception throwing an exception here will cause it to be logged and also sent to [.onError]
+     */
+    @Throws(Exception::class)
+    override fun onMessage(event: String?, messageEvent: MessageEvent?) {
+        Timber.d("MESSAGE")
```
Comment (ajwootto) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt:43]: these also seem like temporary logs to me, or ones that could be more descriptive

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    /**
+     * EventSource calls this method when it has received a new event from the stream.
+     * @param event the event name, from the `event:` line in the stream
+     * @param messageEvent a [MessageEvent] object containing all the other event properties
+     * @throws Exception throwing an exception here will cause it to be logged and also sent to [.onError]
+     */
+    @Throws(Exception::class)
+    override fun onMessage(event: String?, messageEvent: MessageEvent?) {
```
Comment (ajwootto) [inline android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt:42]: how does the client react to new messages received here? you should wire that up here unless someone is doing it in a later PR

Comment (elliotCamblor) [reply android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt:42]: That would be done in the ticket thats for refetching the config when a new message is received

File: android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt (hunk 11 above)
Comment (elliotCamblor) [reply android-client-sdk/src/main/java/com/devcycle/sdk/android/eventsource/EventHandler.kt:43]: yeah its just a placeholder, but like you said before I should make sure that code on main could be released at anytime, so ill remove it

**Review by ajwootto — APPROVED** [review]

**Review by jsalaber — APPROVED** [review]

**Pull request merged**